### PR TITLE
Pool more temporary values while doing camel case matching.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Tagging @ManishJayaswal 

Another PR which is just about producing less garbage when we do large scale pattern matching (i.e. for NavigateTo).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @grabalon

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs [hunk 1]
```diff
@@ -50,25 +52,31 @@ public AllLowerCamelCaseMatcher(string candidate, bool includeMatchedSpans, Stri
...
-                            matchedSpans: _includeMatchedSpans ? new List<TextSpan>() : null);
+                    var weight = contiguous == false ? 0 : CamelCaseContiguousBonus;
+                    var matchedSpansInReverse = _includeMatchedSpans ? ArrayBuilder<TextSpan>.GetInstance() : null;
+
+                    return (weight, matchedSpansInReverse);
                 }
 
                 var bestWeight = default(int?);
-                var bestMatchedSpans = default(List<TextSpan>);
+                var bestMatchedSpansInReverse = default(ArrayBuilder<TextSpan>);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: 😦 Not a fan of using `default(T)` for cases where `T` is known to be a reference type. It confuses the syntax with cases like `default(ImmutableArray<int>)` which produces a non-null value with "interesting" semantics. The above line could be improved by writing it as the following, even though it no longer uses `var`.

```csharp
ArrayBuilder<TextSpan> bestMatchedSpansInReverse = null;
```

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs [hunk 2]
```diff
@@ -102,13 +110,14 @@ public AllLowerCamelCaseMatcher(string candidate, bool includeMatchedSpans, Stri
                             contiguous = true;
                         }
 
-                        var (weight, matchedSpans) = TryConsumePatternOrMatchNextHump(
+                        var (weight, matchedSpansInReverse) = TryConsumePatternOrMatchNextHump(
                             patternIndex, humpIndex, contiguous.Value);
                         if (weight == null)
                         {
+                            Debug.Assert(matchedSpansInReverse == null);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: :memo: During my first-pass read, my instinct based on the variable declaration form used above was this check would always fail. It turns out this is not the case, but it goes to show why the implications of `default(T)` are an important factor in a decision to use it.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs [hunk 3]
```diff
@@ -118,27 +127,40 @@ public AllLowerCamelCaseMatcher(string candidate, bool includeMatchedSpans, Stri
                             // We found a path that allowed us to match everything contiguously
                             // from the beginning.  This is the best match possible.  So we can
                             // just stop now and return this result.
-                            return (weight, matchedSpans);
+                            //
+                            // Return any previous best results back to the pool.
+                            bestMatchedSpansInReverse?.Free();
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: :bulb: If you move the containing block below the next block, you can remove the special handling this call to `Free`, and also just keep the original comment. If you do this, the return statement would be changed to:

```csharp
return (weight, bestMatchedSpansInReverse);
```

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs [hunk 4]
```diff
@@ -184,23 +207,36 @@ public AllLowerCamelCaseMatcher(string candidate, bool includeMatchedSpans, Stri
                     {
                         // We found a path that allowed us to match everything contiguously
                         // from the beginning.  This is the best match possible.  So we can
-                        // just stop now and return thie result.
-                        matchedSpans?.Insert(0, candidateMatchSpan);
-                        return (weight, matchedSpans);
+                        // just stop now and return this result.
+                        //
+                        // Return any previous best results back to the pool.
+                        matchedSpansInReverse?.Add(candidateMatchSpan);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: :bulb: This method could also benefit from block reordering like I mentioned above.

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 

This is work to reduce unnecessary allocations in NavigateTo.  Thanks!

**Comment by MattGertz** [comment]
Approved pending tests.

**Comment by CyrusNajmabadi** [comment]
@sharwell Will look into your suggestions!

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: Nice flow clarification.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.AllLowerCamelCaseMatcher.cs]: Done.

**Pull request merged**